src/mito2/src/sst/index/fulltext_index/applier.rs

```diff
@@ -101,129 +86,252 @@ impl FulltextIndexApplier {
         &self,
         file_id: FileId,
         file_size_hint: Option<u64>,
-    ) -> Result<BTreeSet<RowId>> {
+    ) -> Result<Option<BTreeSet<RowId>>> {
         let _timer = INDEX_APPLY_ELAPSED
             .with_label_values(&[TYPE_FULLTEXT_INDEX])
             .start_timer();
 
-        let mut inited = false;
-        let mut row_ids = BTreeSet::new();
-
-        'outer: for (column_id, request) in &self.requests {
-            let dir = self
-                .index_dir_path(file_id, *column_id, file_size_hint)
-                .await?;
-            let path = match &dir {
-                Some(dir) => dir.path(),
-                None => {
-                    // Return empty set if the index not found.
-                    return Ok(BTreeSet::new());
-                }
-            };
+        let mut row_ids: Option<BTreeSet<RowId>> = None;
+        for (column_id, request) in &self.requests {
+            if request.queries.is_empty() {
+                continue;
+            }
 
-            let searcher =
-                TantivyFulltextIndexSearcher::new(path).context(ApplyFulltextIndexSnafu)?;
+            let Some(result) = self
+                .apply_one_column(file_size_hint, file_id, *column_id, request)
+                .await?
+            else {
+                continue;
+            };
 
-            for query in &request.queries {
-                let result = searcher
-                    .search(&query.0)
-                    .await
-                    .context(ApplyFulltextIndexSnafu)?;
+            if row_ids.is_none() {
+                row_ids = Some(result);
+            } else if let Some(ids) = row_ids.as_mut() {
+                ids.retain(|id| result.contains(id));
+            }
```

Comment:
non-blocking: Consider this style. We can also check whether rows_ids is empty in the if-let branch after `retain()`.
```suggestion
            if let Some(ids) = row_ids.as_mut() {
                ids.retain(|id| result.contains(id));
            } else {
                row_ids = Some(result);
            }
```